DDS-1053_elasticsearch_reindex

dmlond opened the pull request:
implements versioned elasticsearch indices aliased to standard elasticsearch indices
    - includes mapping and migration version
    - these are hard-coded into the models
  implements ElasticsearchHandler service that contains reusable elasticsearch methods that were
  formerly only in the rake task.

ElasticsearchHandler implements smart_reindex
    - if the migration version has not changed, will fast reindex data from the old version to the new version
    - if the migration_version has changed, requires a manual, slow reindex

Donmurry committed:
Merge pull request #1110 from Duke-Translational-Bioinformatics/ua_test

Production

dmann committed:
Merge pull request #1111 from Duke-Translational-Bioinformatics/develop

Merging v1.4.13 to ua_test

dmlond committed:
Merge pull request #1114 from Duke-Translational-Bioinformatics/develop

Merge v1.4.14 to ua_test

dmann committed:
Merge remote-tracking branch 'upstream/production' into prod_to_uatest

dmann committed:
Merge pull request #1116 from dmann/prod_to_uatest

Prod to uatest

dmlond committed:
Merge pull request #1115 from Duke-Translational-Bioinformatics/ua_test

Merge v1.4.x to production

dmlond committed:
SearchableModel now expected to implement
  - self.mapping_verion should return a uuid
  - self.migration_version should return a uuid
  - self.versioned_index_name should return "#{self.index_name}.#{self.mapping_version}.#{self.migration_version}"

DataFile, Folder, and Activity meet SearchableModel expectations

dmlond committed:
ensure versioned_index_name matches the index_name, mapping_version, and migration_version instead of parsing with split

dmlond committed:
elasticsearch index names must user lowercase letters

dmlond committed:
Merge branch 'production' of github.com:Duke-Translational-Bioinformatics/duke-data-service into DDS-1053-elasticsearch_reindex

dmann committed:
Remove ES index creation from spec/rails_helper.rb

dmann committed:
Removing all ES indices before every example to determine exactly what requires indices

dmann committed:
Removing all ES indices AFTER every example to determine exactly what requires indices...

dmann committed:
Adding the elasticsearch prep context where needed

dmann committed:
Only remove ES indicies once at the begining of the suite

dmlond committed:
Merge pull request #1 from dmann/DDS-1058-stop_creating_indices_before_all_tests

DDS-1058 stop creating indices before all tests

dmlond committed:
ElasticsearchHandler wraps logic that was previously in the elasticsearch rake tasks, and
adds new features

ElasticsearchHandlerSpec expects ElasticsearchHandler to implement the following correctly:
  - new
  - new(verbose: true)
  - create_indices
    - now creates each index named Model.versioned_index_name, aliased to Model.index_name aliases
  - index_documents
  - drop_indices
    - now drops '_all' indices
  - smart_reindex_indices
    - if index migration_version matches current Model.migration_version, will use fast_reindex
  - start_scroll
  - next_scroll
  - send_batch
  - batch_from_search
  - fast_reindex
    - uses elasticsearch scroll search to fully copy all documents from a source_client, source_index
      to a target_client, target_index
    - source_client and target_client can clients attached to different Elasticsearch clusters

search_shared_examples.rb uses ElasticsearchHandler to drop and create indices correctly

dmlond committed:
uses ElasticsearchHandler for all tasks
ElasticsearchRakeSpec mocks many ElasticsearchHandler functions for more efficient test
  now that ElasticsearchHandlerSpec tests all of these functions more fully

dmlond committed:
uses ElasticsearchHandler #drop_indices

dmlond committed:
Activity is a SearchableModel, needs to be in FolderFilesResponse.indexed_models

dmlond committed:
All SearchableModels should be in FolderFilesResponse.indexed_models

dmlond committed:
Actiivty is now an indexed_model

dmann commented on app/models/activity.rb:
```diff
@@ -2,7 +2,25 @@ class Activity < ActiveRecord::Base
+  @@migration_version = '429d08e4-622d-456a-a87d-84c288857320'
+
+  # change this variable to a new uuid (lowercase letters!) any time the mappings below change
+  @@mapping_version = 'bab2183a-5ed5-4cd3-b05e-916e20817dd7'
```
Not sure we need to use class variables for these versions. I think we could just place the strings directly in the methods.

dmann commented on spec/support/concerns/searchable_model_shared_examples.rb:
```diff
@@ -5,6 +5,19 @@
+  describe 'elasticsearch migrations' do
+    it { expect(described_class).to respond_to(:index_name) }
+    it { expect(described_class).to respond_to(:mapping_version) }
+    it { expect(described_class).to respond_to(:migration_version) }
```
Should test that the *_version methods don't raise NotImplementedError.

dmlond replied:
the test that includes line 16 will fail if a NotImplementedError is raised

dmlond committed:
ElasticsearhHandlerSpec expects ElasticsearchHandler
  - create_indices to allow an optional client (default Elasticsearch::Model.client)
    and create indices in that client
  - drop_indices to allow an optional client (default Elasticsearch::Model.client)
    and drop indices in that client

dmlond committed:
elasticsearch:index:create and elasticsearch:index:drop allow an optional
ENV[TARGET_URL] which forces the indices to be added to/dropped from the
elasticsearch cluster at the supplied URL.

dmann commented on app/services/elasticsearch_handler.rb:
```diff
@@ -0,0 +1,181 @@
+    client.indices.delete index: '_all'
+  end
+
+  def smart_reindex_indices
```
This method is 42 lines long, with `if` statements nested 4 levels deep. It /should/ be refactored, but in the interest of time, we can pair tomorrow and walk through the code together.

dmann reviewed (commented):
I'm not finished reviewing. There's a lot here and I think it would be more efficient to go through it together.

dmlond committed:
ElasticsearchHandler #create_indices does not drop all indices before running, and
  only creates an index for the versioned_index_name if it does not already exist.

Code Review Changes:
Activity, DataFile, Folder:
  removed @@migration_version, and @@mapping_version and moved strings into methods

Added comments to ElasticsearchHandler large, complex methods
Refactored fast_reindex for DRYness

dmann reviewed (approved):
👍
